### ISSUE TITLE
Add URI functionality in a library

### ISSUE THREAD

**Issue by CrowdHailer** [gleam-lang/suggestions#63]
`import gleam/uri`

There is functionality to parse uri strings in the erlang core. http://erlang.org/doc/man/uri_string.html
I would suggest this module could start being useful with just a single function `parse`. This is useful in all sorts of places, for example parsing a database_url.

I think it's worth wrapping this in a core lib because the function in the `uri_string` module, returns a map with optional keys and heterogeneous values (mostly strings but port is an integer).
Because URI's don't have to have all the parts I'm not sure what the most convenient return type is.
I think it is `Result(Uri)` where

```
type Uri {
Uri(scheme: Option(String), userinfo: Option(String), ..etc)
}
```

**Comment by lpil** [comment]
As with other useful types (HTTP types, dates, etc) I'd like to see the implemented as a library first and then once we are happy with the design it can be prompted into the stdlib. I would like to avoid churn in the stdlib where possible.

**Comment by CrowdHailer** [comment]
I'm not (yet) interested in maintaining lots of small libraries. However here is my implementation of a `Uri` module from midas, which might be useful for anyone else coming across this issue.

https://github.com/midas-framework/midas/tree/master/src/uri

**Comment by CrowdHailer** [comment]
I've changed my mind, I don't think there are any hard design choices to be made in a URI library. Most of the requisite terms are specified in RFC. Therefore it would not be that risky to put a URI library in stdlib.

Also when you consider using URI's for configuration, e.g a database url in a postgres client then there are quite a few occasions where a url.parse function in stdlib would be useful.

I've implemented just the core functionality of a URI library here (basically the minimal amount to be useful), and documented it.
https://github.com/midas-framework/midas/blob/master/src/gleam/uri.gleam

Looking at whats in the Elixir/erlang standard library I think adding this core of a uri library would be a safe addition to standard lib. Finally we're still pre 1.0 so changes some changes are allowed.

I'm happy to open a PR on stdlib if it would be acceptable.
